### PR TITLE
feat(WebWorkers): Add WebSocket MessageBuses for debugging apps

### PR DESCRIPTION
**Pull request by jteplitz**

### CODE REVIEW

File: modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart [hunk 1]
```diff
@@ -0,0 +1,215 @@
+library angular2.src.web_workers.debug_tools.multi_client_server_message_bus;
+
+import "package:angular2/src/web_workers/shared/message_bus.dart"
+    show MessageBus, MessageBusSink, MessageBusSource;
+import 'dart:io';
+import 'dart:convert';
+import 'dart:async';
+import "package:angular2/src/core/facade/async.dart" show EventEmitter;
+import "package:angular2/src/web_workers/shared/messaging_api.dart";
```
Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart]: use `'` everywhere for consistency

File: modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+class WebSocketWrapper {
+  WebSocket socket;
+  Stream stream;
+  int numResultsReceived = 0;
+  bool isPrimary = false;
+  bool caughtUp = false;
+  List<String> messageHistory;
+  List<int> resultMarkers;
```
Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart]: do all the fields above need to be publicly writable?

File: modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+  final JsonDecoder decoder = new JsonDecoder();
```
Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart]: `JsonDecoder` is stateless. You can make it `static final`

Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/multi_client_server_message_bus.dart]: do all fields need to be public at all?

File: modules/angular2/src/web_workers/debug_tools/web_socket_message_bus.dart [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  }
+}
+
+class WebSocketMessageBusSource implements MessageBusSource {
+  final JsonDecoder decoder = new JsonDecoder();
+  final Map<String, EventEmitter> _channels = new Map<String, EventEmitter>();
+
+  WebSocketMessageBusSource(WebSocket webSocket) {
+    webSocket.onMessage.listen((MessageEvent encodedMessage) {
+      //print("Got message " + encodedMessage.data);
```
Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/web_socket_message_bus.dart]: remove commented out code

File: modules/angular2/src/web_workers/debug_tools/web_socket_message_bus.dart [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  EventEmitter to(String channel) {
+    return sink.to(channel);
+  }
+}
+
+class WebSocketMessageBusSink implements MessageBusSink {
+  final WebSocket _webSocket;
+  final JsonEncoder encoder = new JsonEncoder();
```
Comment (yjbanov) [inline modules/angular2/src/web_workers/debug_tools/web_socket_message_bus.dart]: If you are not customizing JSON encoding/decoding reuse the global instance `JSON` available in the `dart:convert` package:

```dart
import 'dart:convert' show JSON;

JSON.encode(...);
JSON.decode(...);
```

File: modules/angular2/test/web_workers/debug_tools/message_bus_common.dart [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+import "dart:convert";
+import "package:angular2/src/web_workers/shared/message_bus.dart";
+import "package:angular2/test_lib.dart"
+    show AsyncTestCompleter, expect, SpyObject;
+
+var MESSAGE = const {'test': 10};
+const CHANNEL = "TEST_CHANNEL";
+
+void expectSinkSendsEncodedJSON(SpyObject socket, MessageBusSink sink,
```
Comment (yjbanov) [inline modules/angular2/test/web_workers/debug_tools/message_bus_common.dart]: Abbreviations in variable/function/method/class names, such as `JSON`, should be camel-cased, e.g. 'Json' (constants use all caps).

File: modules/angular2/test/web_workers/debug_tools/multi_client_server_message_bus.server.spec.dart [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    const CHANNEL = "TEST_CHANNEL";
+    var MESSAGE = const {'test': 10};
+
+    beforeEach(() {
+      messageHistory.clear();
+      resultMarkers.clear();
+    });
+
+    it(
+        "should send messages to all connected clients",
```
Comment (yjbanov) [inline modules/angular2/test/web_workers/debug_tools/multi_client_server_message_bus.server.spec.dart]: I wonder if `dartfmt` agrees with this styling

Comment (jteplitz602) [inline modules/angular2/test/web_workers/debug_tools/multi_client_server_message_bus.server.spec.dart]: ```dartfmt``` actually created that styling :smiley:

**Pull request closed without merging**